bite-size-notes/week-6-b-free-final-time-PMP.md

```diff
@@ -0,0 +1,231 @@
+$$
+\begin{equation}
+\nabla_T \Psi(z(T)) + H(T) = 0
+\end{equation}
+$$
+
+### Transversality Conditions
+
+For free final time, the transversality conditions state that at the optimal terminal time $T$, the Hamiltonian $H$ should satisfy
+$$H\big(z(T),u(T),\lambda(T),T\big) = 0.$$ This implies that no further performance can be achieved by changing $T$.
+
+For similar reasons, all co-states should be zero at the optimal terminal time $T$.
+$$\lambda(T) = 0$$
+
+
+## Example: 2D Unicycle
+
+Now let's look at a standard optimal control problem, but this time with terminal time being an optimization variable.
+
+### Analytical solution
+Let's revisit a common example: the 2D unicycle. We seek to reach a goal state and minimize both effort and time doing so.
+* state variable $z$
+  * positions $x$, $y$
+  * velocity $v$
+  * angle $\theta$
+* control input $u$
+  * acceleration $a$
+  * angular velocity $\omega$
+* state variable $z_g$
+  * goal positions $x_g$, $y_g$
+  * goal velocity $v_g$
+  * goal angle $\theta_g$
+* co-state $\lambda$
+* Lagrange multiplier $\mu$
+* maximum acceleration $a_{\text{max}}$
+
+Then the state and dynamics vectors, still functions of time, are:
+
+$$
+z = \begin{bmatrix}
+  x \\
+  y \\
+  v \\
+  \theta
+\end{bmatrix}, \qquad
+\text{and}
+\qquad \dot{z} = \begin{bmatrix}
+  v \cos \theta \\
+  v \sin \theta \\
+  a \\
+  \omega
+\end{bmatrix}
+$$
+
+Let the Lagrangian $L(z(t), u(t), t\big) = a^2 + \omega^2 + \|x-x_g\|^2 + \|y-y_g\|^2$ and the terminal cost be $\Psi(T) = \|z(T) - z_g\|^2$. Include a constraint on the control: $a \le a_{\text{max}}$.
+
+The cost functional can be written as:
+$$
+\begin{align*} J[z(t),u(t),t]
+&= \int_0^T L\big(z(t), u(t), t\big) \,dt + \Psi(T)\\
+&= \int_0^T a^2 + \omega^2 + \|x-x_g\|^2 + \|y-y_g\|^2 \,dt + \|z(T) - z_g\|^2\\
+\end{align*}
+$$
+
+Augment the Lagrangian with the control constraint with a Lagrange multiplier of $\mu$.
+$$L(z(t), u(t), t\big) = a^2 + \omega^2 + \|x-x_g\|^2 + \|y-y_g\|^2 + \mu (a_{\text{max}} - a) $$
+
+Then the Hamiltonian augments the state dynamics to the Lagrangian.
+$$H(z(t), u(t), t\big) = \lambda(t)^\top \cdot \dot{z} + a^2 + \omega^2 + \|x-x_g\|^2 + \|y-y_g\|^2 + \mu (a_{\text{max}} - a) $$
+
+Taking the second derivative of the Hamiltonian, 
+$$
+\frac{\partial H}{\partial z} = \begin{bmatrix} 2 & 0 & 0 & 0 \\
+0 & 2 & 0 & 0 \\
+0 & 0 & 0 & 0 \\
+0 & 0 & 0 & 0 \end{bmatrix}
+$$
+
+By Pontryagin’s minimum principle and transversality conditions, we must satisfy these conditions to be optimal.
+$$
+H(z^*(t), u^*(t), \lambda^*(t), t) \leq H(z(t), u(t), \lambda(t), t) \text{  for  } t \in [0,T]
+$$
+
+$$
+-\dot{\lambda}(t) = \lambda(t)^\top \cdot \nabla_z \dot{z}(z^*(t), u^*(t)) + \nabla_z L(z^*(t), u^*(t)) 
+$$
+
+$$
+\lambda(T)^\top = \nabla_z \Psi(z(T))
+$$
+
+$$
+\nabla_T \Psi(z(T)) + H(T) = 0
+$$
+
+$$H\big(z(T),u(T),\lambda(T),T\big) = 0$$
```

Comment:
N.B.: Check comment on line 127: if you impose two conditions on $\lambda(T)$ and on $H(t)$ that are different is the problem solvable? Thus one condition should specialize to the other for particular choice of $\Psi$

Reply:
Agree with that insight, I misunderstood an example from a source, it's been fixed.